File: DSA/src/Questions/BinarySearch/PeakElement.java
package Questions.BinarySearch;

import java.util.ArrayList;
import java.util.Arrays;
/*
 https://leetcode.com/problems/find-peak-element/description/

If arr[mid] is greater than both its neighbors (or if it is at the boundary and greater than its single neighbor),
 it is a peak element. 
 if ((mid == 0 || arr.get(mid - 1) < arr.get(mid)) && (mid == n - 1 || arr.get(mid) > arr.get(mid + 1))) {
                return mid;
            }

If arr[mid] is less than its left neighbor, move the search range to the left half.

If arr[mid] is less than its right neighbor, move the search range to the right half.

 */
public class PeakElement {

    public static int findPeakElement(ArrayList<Integer> arr) {
        int n = arr.size(); // Size of the array

        int low = 0, high = n - 1;
        while (low <= high) {
            int mid = low + (high - low) / 2;

            // Check if arr[mid] is the peak:
            if ((mid == 0 || arr.get(mid - 1) < arr.get(mid)) && (mid == n - 1 || arr.get(mid) > arr.get(mid + 1))) {
                return mid;
            }

            // If we are in the left and arr[mid] is decreasing, move right:
            if (mid > 0 && arr.get(mid) < arr.get(mid - 1)) {
                high = mid - 1;
            } else {
                // If we are in the right or arr[mid] is increasing, move left:
                low = mid + 1;
            }
        }
        // If no peak is found, return -1:
        return -1;
    }

    public static void main(String[] args) {
        ArrayList<Integer> arr = new ArrayList<>(Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 5, 1));
        int ans = findPeakElement(arr);
        if (ans != -1) {
            System.out.println("The peak is at index: " + ans);
        } else {
            System.out.println("No peak found in the array.");
        }
    }
}

File: DSA/Python/Questions/BinarySearch/PeakElement.py
"""
# Question: Find Peak Element
# Link: https://leetcode.com/problems/find-peak-element/

# Problem Statement:
# A peak element is an element that is strictly greater than its neighbors.
# Given an integer array nums, find a peak element, and return its index.
# If the array contains multiple peaks, return any of the peaks' index.

# Example:
# Input: nums = [1,2,3,1]
# Output: 2
# Explanation: 3 is a peak element and your function should return the index number 2.
"""

from typing import List


class PeakElement:
    def find_peak_binary(self, nums: List[int]) -> int:
        """
        Binary search approach
        Time Complexity: O(log n)
        Space Complexity: O(1)
        """
        left, right = 0, len(nums) - 1

        while left < right:
            mid = (left + right) >> 1
            if nums[mid] > nums[mid + 1]:
                right = mid
            else:
                left = mid + 1

        return left

    def find_peak_linear(self, nums: List[int]) -> int:
        """
        Linear search approach
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        for i in range(len(nums) - 1):
            if nums[i] > nums[i + 1]:
                return i
        return len(nums) - 1

    def find_peak_recursive(self, nums: List[int]) -> int:
        """
        Recursive approach
        Time Complexity: O(log n)
        Space Complexity: O(log n)
        """

        def find_peak_helper(left: int, right: int) -> int:
            if left == right:
                return left

            mid = (left + right) >> 1
            if nums[mid] > nums[mid + 1]:
                return find_peak_helper(left, mid)
            return find_peak_helper(mid + 1, right)

        return find_peak_helper(0, len(nums) - 1)


def main():
    test_cases = [
        {"nums": [1, 2, 3, 1], "expected_values": {2}},
        {"nums": [1, 2, 1, 3, 5, 6, 4], "expected_values": {1, 5}},
        {"nums": [1], "expected_values": {0}},
    ]

    solution = PeakElement()
    for i, test in enumerate(test_cases):
        print(f"\nTest Case {i + 1}:")
        print(f"Input: {test['nums']}")

        result_binary = solution.find_peak_binary(test["nums"])
        result_linear = solution.find_peak_linear(test["nums"])
        result_recursive = solution.find_peak_recursive(test["nums"])

        print(f"Binary Search Result: {result_binary}")
        print(f"Linear Search Result: {result_linear}")
        print(f"Recursive Result: {result_recursive}")

        # Verify results are valid peaks
        assert result_binary in test["expected_values"]
        assert result_linear in test["expected_values"]
        assert result_recursive in test["expected_values"]


if __name__ == "__main__":
    main()
